# Introduction to scikit-learn (sklearn)
Demonstrate how scikit-learn works

0. An end-to-end scikit-learn workflow
1. Getting the data ready
2. Choose the right estimator/algorithm for our problems
3. Fit the model/algorithm and use it to make predictions on our data
4. Evaluating a model
5. Improve a model
6. Save and load a trained model
7. Putting it all together!

## 0. An end-to-end scikit-learn workflow

## 1. Getting the data ready

In [1]:
import pandas as pd
import numpy as np
heart_disease = pd.read_csv("../heart-disease.csv")
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [2]:
# Create x (features matrix)
x = heart_disease.drop("target", axis=1)

# Create y (labels)
y = heart_disease["target"]

## 2. Choose the right model and hyperparameters

In [3]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()

# Keep the default hyperparameters
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

## 3. Fit the model to the training data

In [4]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [5]:
clf.fit(x_train, y_train);

In [6]:
y_preds = clf.predict(x_test)

In [7]:
y_preds

array([1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0], dtype=int64)

In [8]:
y_test

130    1
220    0
163    1
299    0
288    0
      ..
54     1
281    0
224    0
226    0
279    0
Name: target, Length: 61, dtype: int64

## 4. Evaluate the model on the training data and test data

In [9]:
clf.score(x_train, y_train)

1.0

### The model scored 1 out of 1 on the test data, which is very good. 

In [10]:
clf.score(x_test, y_test)

0.8852459016393442

### The model scored correct on 83% of the test data (data that the model has not seen before)

In [11]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       1.00      0.77      0.87        30
           1       0.82      1.00      0.90        31

    accuracy                           0.89        61
   macro avg       0.91      0.88      0.88        61
weighted avg       0.91      0.89      0.88        61



In [12]:
confusion_matrix(y_test, y_preds)

array([[23,  7],
       [ 0, 31]], dtype=int64)

In [13]:
accuracy_score(y_test, y_preds)

0.8852459016393442

## 5. Improve the model

In [14]:
# Try differente amount of n_estimators
np.random.seed(42)
for i in range(10, 100, 10): 
    print(f"Trying model with {i} estimators...")
    clf = RandomForestClassifier(n_estimators=i).fit(x_train, y_train)
    print(f"Model accuracy on test set: {clf.score(x_test, y_test) * 100:.2f}%")
    print("")

Trying model with 10 estimators...
Model accuracy on test set: 88.52%

Trying model with 20 estimators...
Model accuracy on test set: 86.89%

Trying model with 30 estimators...
Model accuracy on test set: 88.52%

Trying model with 40 estimators...
Model accuracy on test set: 86.89%

Trying model with 50 estimators...
Model accuracy on test set: 86.89%

Trying model with 60 estimators...
Model accuracy on test set: 83.61%

Trying model with 70 estimators...
Model accuracy on test set: 83.61%

Trying model with 80 estimators...
Model accuracy on test set: 85.25%

Trying model with 90 estimators...
Model accuracy on test set: 86.89%



### Choosing more estimators does not improve the model from the original. 83.61% is the best we can get using 10 estimators. 

In [15]:
clf = RandomForestClassifier(n_estimators=10).fit(x_train, y_train)

## 6. Save the model and load it

In [16]:
import pickle

pickle.dump(clf, open("random_forest_model_1.pkl", "wb")) # wb = write binary

In [17]:
loaded_model = pickle.load(open("random_forest_model_1.pkl", "rb")) # rb = read binary
loaded_model.score(x_test, y_test)

0.8360655737704918

## 1. Getting our data ready to be used with machine learning

**Clean Data -> Transform Data -> Reduce data**

Three main things we have to do
* Split the data into features and labels (usually 'X' and 'y')
* Filling (also called computing) or disregarding missing values
* Converting non-numerical values to numerical values (also called feature encoding9

In [18]:
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [19]:
X = heart_disease.drop("target", axis=1)
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [20]:
y = heart_disease["target"]
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

## Split the data into training and test set

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [22]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((242, 13), (61, 13), (242,), (61,))

## Make sure it's all numerical 

* Using car_sales to practice this since heart_disease is already all numerical

In [23]:
car_sales = pd.read_csv("../car-sales-extended.csv")
car_sales.head()

Make Colour  Odometer (KM)  Doors  Price
0   Honda  White          35431      4  15323
1     BMW   Blue         192714      5  19943
2   Honda  White          84714      4  28343
3  Toyota  White         154365      4  13434
4  Nissan   Blue         181577      3  14043

In [24]:
len(car_sales)

1000

In [25]:
car_sales.dtypes

Make             object
Colour           object
Odometer (KM)     int64
Doors             int64
Price             int64
dtype: object

## We need to convert these to numbers, but let's try anyway and see what happens

In [26]:
# split into X/y

X = car_sales.drop("Price", axis=1)
y = car_sales["Price"]

# split into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [27]:
# Build machine learning model
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

ValueError: could not convert string to float: 'Toyota'

### We get an type error because when running the machine learning model, we cannot convert the string to a float/int. 
* This is why we have to convert values to numbers when doing machine learning since the computer does not know what a car brand name is

## Let's convert them into numbers using sklearn

## OneHotEncoder 

turns categorical values into numerical values. In this case, colour is the column we'd like to transform 

**(Car: 1337, Colour: Red, Kilometers: 1008) => (0: 1337, 1, 0, 0, 1008)**

Adding two extra columns of 0 to demonstrate RGB

In [28]:
# Turn the categories into numbers
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"] # Why doors, it's numerical? It can be categorical, so a car with 4 doors is one category, 5 is another, 3 is another
one_hot = OneHotEncoder() 
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder="passthrough")

transformed_X = transformer.fit_transform(X)
transformed_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        1.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.48360e+05]])

In [29]:
X.head()

Make Colour  Odometer (KM)  Doors
0   Honda  White          35431      4
1     BMW   Blue         192714      5
2   Honda  White          84714      4
3  Toyota  White         154365      4
4  Nissan   Blue         181577      3

In [30]:
pd.DataFrame(transformed_X).head()

0    1    2    3    4    5    6    7    8    9    10   11        12
0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   35431.0
1  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  192714.0
2  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   84714.0
3  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  154365.0
4  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  181577.0

In [31]:
# This is another way using pandas instead of OneHotEncoder
dummies = pd.get_dummies(car_sales[["Make", "Colour", "Doors"]])
dummies

Doors  Make_BMW  Make_Honda  Make_Nissan  Make_Toyota  Colour_Black  \
0        4         0           1            0            0             0   
1        5         1           0            0            0             0   
2        4         0           1            0            0             0   
3        4         0           0            0            1             0   
4        3         0           0            1            0             0   
..     ...       ...         ...          ...          ...           ...   
995      4         0           0            0            1             1   
996      3         0           0            1            0             0   
997      4         0           0            1            0             0   
998      4         0           1            0            0             0   
999      4         0           0            0            1             0   

     Colour_Blue  Colour_Green  Colour_Red  Colour_White  
0              0             0           0             1  
1              1             0           0             0  
2              0             0           0             1  
3              0             0           0             1  
4              1             0           0             0  
..           ...           ...         ...           ...  
995            0             0           0             0  
996            0             0           0             1  
997            1             0           0             0  
998            0             0           0             1  
999            1             0           0             0  

[1000 rows x 10 columns]

In [32]:
# Let's refit the model
np.random.seed(42)
X_train, X_test, y_train, y_test = train_test_split(transformed_X, y, test_size=0.2)
model.fit(X_train, y_train)

RandomForestRegressor()

In [33]:
model.score(X_test, y_test)

0.3235867221569877

### OneHotEncoder can now handle NaN/None values

#### What if there were missing values?

1. Fill them with some value (also known as imputation).
2. Remove the samples with missing data altogether

In [34]:
car_sales_missing = pd.read_csv("../car-sales-extended-missing-data.csv")
car_sales_missing.head()

Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0

In [35]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

### Let's try and convert data to numbers

In [36]:
# Create X and Y
X = car_sales_missing.drop("Price", axis=1)
y = car_sales_missing["Price"]

In [37]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]
one_hot = OneHotEncoder(sparse_output=False)
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder="passthrough")

transformed_X = transformer.fit_transform(X)
pd.DataFrame(transformed_X)

0    1    2    3    4    5    6    7    8    9    10   11   12   13  \
0    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
1    1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
2    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
3    0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
4    0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
995  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
996  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0   
997  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
998  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
999  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   

      14        15  
0    0.0   35431.0  
1    0.0  192714.0  
2    0.0   84714.0  
3    0.0  154365.0  
4    0.0  181577.0  
..   ...       ...  
995  0.0   35820.0  
996  0.0  155144.0  
997  0.0   66604.0  
998  0.0  215883.0  
999  0.0  248360.0  

[1000 rows x 16 columns]

### Option 1: Fill missing data with pandas

In [38]:
# Fill the "Make" column
car_sales_missing["Make"].fillna(value="missing", inplace=True)

# Fill the "Colour" column
car_sales_missing["Colour"].fillna(value="missing", inplace=True)

# Fill the "Odometer (KM)" column
car_sales_missing["Odometer (KM)"].fillna(value=car_sales_missing["Odometer (KM)"].mean(), inplace=True)

# Fill the "Doors" column
car_sales_missing["Doors"].fillna(value=4, inplace=True)

print(car_sales_missing.isna().sum())
car_sales_missing

Make              0
Colour            0
Odometer (KM)     0
Doors             0
Price            50
dtype: int64


Make Colour  Odometer (KM)  Doors    Price
0      Honda  White        35431.0    4.0  15323.0
1        BMW   Blue       192714.0    5.0  19943.0
2      Honda  White        84714.0    4.0  28343.0
3     Toyota  White       154365.0    4.0  13434.0
4     Nissan   Blue       181577.0    3.0  14043.0
..       ...    ...            ...    ...      ...
995   Toyota  Black        35820.0    4.0  32042.0
996  missing  White       155144.0    3.0   5716.0
997   Nissan   Blue        66604.0    4.0  31570.0
998    Honda  White       215883.0    4.0   4001.0
999   Toyota   Blue       248360.0    4.0  12732.0

[1000 rows x 5 columns]

In [39]:
# Remove rows with missing Price value, hard to train to predict something that's missing a target value (y)
car_sales_missing.dropna(inplace=True)

In [40]:
car_sales_missing.isna().sum()

Make             0
Colour           0
Odometer (KM)    0
Doors            0
Price            0
dtype: int64

In [41]:
len(car_sales_missing)

950

In [42]:
# Create X and Y
X = car_sales_missing.drop("Price", axis=1)
y = car_sales_missing["Price"]

In [43]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]
one_hot = OneHotEncoder(sparse_output=False)
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder="passthrough")

transformed_X = transformer.fit_transform(X)
pd.DataFrame(transformed_X)

0    1    2    3    4    5    6    7    8    9    10   11   12   13  \
0    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
1    1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
2    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
3    0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
4    0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
945  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
946  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0   
947  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
948  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
949  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   

           14  
0     35431.0  
1    192714.0  
2     84714.0  
3    154365.0  
4    181577.0  
..        ...  
945   35820.0  
946  155144.0  
947   66604.0  
948  215883.0  
949  248360.0  

[950 rows x 15 columns]

### Feature scaling

Extension: Feature Scaling
Once your data is all in numerical format, there's one more transformation you'll probably want to do to it.

It's called Feature Scaling.

In other words, making sure all of your numerical data is on the same scale.

For example, say you were trying to predict the sale price of cars and the number of kilometres on their odometers varies from 6,000 to 345,000 but the median previous repair cost varies from 100 to 1,700. A machine learning algorithm may have trouble finding patterns in these wide-ranging variables.

To fix this, there are two main types of feature scaling.

Normalization (also called min-max scaling) - This rescales all the numerical values to between 0 and 1, with the lowest value being close to 0 and the highest previous value being close to 1. Scikit-Learn provides functionality for this in the MinMaxScalar class.

Standardization - This subtracts the mean value from all of the features (so the resulting features have 0 mean). It then scales the features to unit variance (by dividing the feature by the standard deviation). Scikit-Learn provides functionality for this in the StandardScalar class.

A couple of things to note.

Feature scaling usually isn't required for your target variable.

Feature scaling is usually not required with tree-based models (e.g. Random Forest) since they can handle varying features.

Extra reading

For further information on this topic, I'd suggest the following resources.

Feature Scaling - why is it required? by Rahul Saini

Feature Scaling with Scikit-Learn by Ben Alex Keen

Feature Scaling for Machine Learning: Understanding the Difference Between Normalization vs. Standardization by Aniruddha Bhandari

### Option 2: Fill missing value with scikit-learn

In [44]:
# Read the csv again since we have modified the previous one inplace
car_sales_missing = pd.read_csv("../car-sales-extended-missing-data.csv")
car_sales_missing.head()

Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0

In [45]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

In [46]:
# Drop the rows with no price
car_sales_missing.dropna(subset=["Price"], inplace=True)
car_sales_missing.isna().sum()

Make             47
Colour           46
Odometer (KM)    48
Doors            47
Price             0
dtype: int64

In [47]:
# Split into X and y
X = car_sales_missing.drop("Price", axis=1)
y = car_sales_missing["Price"]

In [48]:
from sklearn.model_selection import train_test_split
# Split into train and test data before filling the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [49]:
# Fill missing values with scikit-learn
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer

# Fill categorical values with 'missing' and numerical values with mean
cat_imputer = SimpleImputer(strategy="constant", fill_value="missing") # Run over the categorical values, keep the strategy constant
door_imputer = SimpleImputer(strategy="constant", fill_value=4)
num_imputer = SimpleImputer(strategy="mean")

cat_features = ["Make", "Colour"]
door_feature = ["Doors"]
num_features = ["Odometer (KM)"]

# Create an imputer (something that fills missing data)
imputer = ColumnTransformer([
    ("cat_imputer", cat_imputer, cat_features),
    ("door_imputer", door_imputer, door_feature),
    ("num_imputer", num_imputer, num_features)
])

# Transform the data
filled_X_train = imputer.fit_transform(X_train)
filled_X_test = imputer.transform(X_test)

filled_X_train

array([['Toyota', 'Black', 4.0, 86696.0],
       ['Toyota', 'Black', 4.0, 20714.0],
       ['Toyota', 'White', 4.0, 134415.0],
       ...,
       ['Honda', 'Blue', 4.0, 224900.0],
       ['missing', 'White', 3.0, 19482.0],
       ['Honda', 'Blue', 4.0, 199833.0]], dtype=object)

In [50]:
car_sales_filled_train = pd.DataFrame(filled_X_train, columns=["Make", "Colour", "Doors", "Odometer (KM)"])
car_sales_filled_test = pd.DataFrame(filled_X_test, columns=["Make", "Colour", "Doors", "Odometer (KM)"])
car_sales_filled_train, car_sales_filled_test

(        Make Colour Doors  Odometer (KM)
 0     Toyota  Black   4.0        86696.0
 1     Toyota  Black   4.0        20714.0
 2     Toyota  White   4.0       134415.0
 3     Toyota  Green   4.0  130710.230137
 4      Honda  White   4.0       145850.0
 ..       ...    ...   ...            ...
 755   Nissan   Blue   4.0       111256.0
 756    Honda  White   4.0       246079.0
 757    Honda   Blue   4.0       224900.0
 758  missing  White   3.0        19482.0
 759    Honda   Blue   4.0       199833.0
 
 [760 rows x 4 columns],
         Make   Colour Doors Odometer (KM)
 0     Toyota      Red   4.0      133195.0
 1        BMW    White   5.0       37513.0
 2     Toyota    White   4.0       28809.0
 3     Nissan    White   4.0       82429.0
 4     Nissan    White   4.0      234161.0
 ..       ...      ...   ...           ...
 185  missing    White   4.0      128072.0
 186   Toyota    White   4.0      188338.0
 187   Toyota      Red   4.0      241987.0
 188    Honda  missing   4.0      15058

In [51]:
car_sales_filled_train.isna().sum()

Make             0
Colour           0
Doors            0
Odometer (KM)    0
dtype: int64

In [52]:
filled_X_train

array([['Toyota', 'Black', 4.0, 86696.0],
       ['Toyota', 'Black', 4.0, 20714.0],
       ['Toyota', 'White', 4.0, 134415.0],
       ...,
       ['Honda', 'Blue', 4.0, 224900.0],
       ['missing', 'White', 3.0, 19482.0],
       ['Honda', 'Blue', 4.0, 199833.0]], dtype=object)

In [53]:
# Turn the categories into numbers
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]
one_hot = OneHotEncoder() 
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder="passthrough")

transformed_X_train = transformer.fit_transform(car_sales_filled_train)
transformed_X_test  = transformer.transform(car_sales_filled_test)
transformed_X_train, transformed_X_test

(<760x15 sparse matrix of type '<class 'numpy.float64'>'
 	with 3040 stored elements in Compressed Sparse Row format>,
 <190x15 sparse matrix of type '<class 'numpy.float64'>'
 	with 760 stored elements in Compressed Sparse Row format>)

In [54]:
# Now we got our data as numbers and filled (no missing values)
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor()
model.fit(transformed_X_train, y_train)
model.score(transformed_X_test, y_test)

0.10175069616245203

# 2. Choosing the right model

Some things to note: 

* Sklearn refers to machine learning models, algoritms as estimators
* Classification problem - prediciting a category (heart disease or not, or other types)
    * Sometimes you'll see `clf` (short for classifier) used as a classification estimator
* Regression problem - predicting a number (selling price of a car)

https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html

## 2.1 Picking a machine learning model for a regression problem
Test using california housing dataset

In [55]:
# Get Carlifornia housing dataset
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
housing

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [56]:
housing_df = pd.DataFrame(housing["data"], columns=housing["feature_names"])
housing_df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1        -122.22  
2        -122.24  
3        -122.25  
4        -122.25  
...          ...  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24  

[20640 rows x 8 columns]

In [57]:
housing_df["target (MedHouseVal)"] = housing["target"]
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target (MedHouseVal)  
0    -122.23                 4.526  
1    -122.22                 3.585  
2    -122.24                 3.521  
3    -122.25                 3.413  
4    -122.25                 3.422

In [58]:
# Import algoritm 

# Setup random seed
np.random.seed(42)

# Create the data
X = housing_df.drop("target (MedHouseVal)", axis=1)
y = housing_df["target (MedHouseVal)"]

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Instantiate and fit the model
# After looking at the sklearn model map, we end up at a regression problem for predicting the median house value. Not sure if few features should be important, so let's experiment with some.
from sklearn.linear_model import Ridge
model = Ridge()
model.fit(X_train, y_train)

# Check the score of the model (on the test set)
model.score(X_test, y_test)


0.5758549611440131

### What is ridge didn't work or the scode didn't fit our needs?
- Try a different model

Let's try an ensemble model (An ensemble is combination of smaller models to try and make better predictions than just a single model)

In [59]:
# Import the random RandomForestRegressor model class from the ensemble module
from sklearn.ensemble import RandomForestRegressor

# Setup random seed
np.random.seed(42)

# Create the data

X = housing_df.drop("target (MedHouseVal)", axis=1)
y = housing_df["target (MedHouseVal)"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Create random forest model
model = RandomForestRegressor()
model.fit(X_train, y_train)

# Check the score of the model
model.score(X_test, y_test)

0.8065734772187598

## 2.2 picking a machine learning model for classification problem

In [61]:
heart_disease = pd.read_csv("../heart-disease.csv")
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

### With this dataset, looking at the sklearn map, we should try to use LinearSVC to predict the target. The target is a classification problem

In [81]:
from sklearn.svm import LinearSVC

# X and y
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# split the data into test and train set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

clf = LinearSVC(max_iter=10000)
clf.fit(X_train, y_train)

# Evaluate
clf.score(X_test, y_test)


C:\Users\njesp\Documents\ml\sample_project\env\Lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


0.7868852459016393

In [79]:
heart_disease["target"].value_counts()

1    165
0    138
Name: target, dtype: int64

#### Increasing the max_iter from default 1000 to 10000 makes the classifier better than previous score (68.8%) but it's not really good enough. Let's try another classifier (RandomForestClassifier).

In [82]:
from sklearn.ensemble import RandomForestClassifier

# X and y
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# split the data into test and train set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

clf = RandomForestClassifier()
clf.fit(X_train, y_train)

# Evaluate
clf.score(X_test, y_test)

0.8852459016393442

A little tidbit for modelling in machine learning is:

- If you have structured data (tables or dataframes), use ensemble methods, such as, a Random Forest.
- If you have unstructured data (text, images, audio, things not in tables), use deep learning or transfer learning.

# 3. Fit the model/algorithm on our data and use it to make predictions

### 3.1 Fitting the model to the data

Different names for:
- X = features, features variables, data
- y = labels, targets, target variables

In [83]:
from sklearn.ensemble import RandomForestClassifier

np.random.seed(42)
# X and y
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# split the data into test and train set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

clf = RandomForestClassifier()
# training the machine learning model
clf.fit(X_train, y_train)

# Evaluate
clf.score(X_test, y_test)

0.8524590163934426

In [84]:
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [85]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

### 3.2 make predictions using a machine learning model

2 ways to make predictions:

1. `predict()`
2. `predict_proba()`

In [87]:
# Use a trained model to make predictions
clf.predict(np.array([1, 7, 8, 3, 4])) # this doesn't work, and is expected

C:\Users\njesp\Documents\ml\sample_project\env\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


ValueError: Expected 2D array, got 1D array instead:
array=[1. 7. 8. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [88]:
x_test.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
130   54    0   2       160   201    0        1      163      0      0.0   
220   63    0   0       150   407    0        0      154      0      4.0   
163   38    1   2       138   175    0        1      173      0      0.0   
299   45    1   3       110   264    0        1      132      0      1.2   
288   57    1   0       110   335    0        1      143      1      3.0   

     slope  ca  thal  
130      2   1     2  
220      1   3     3  
163      2   4     2  
299      1   0     3  
288      1   1     3

In [89]:
clf.predict(X_test)

array([0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [92]:
np.array([y_test])

array([[0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
        0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0]], dtype=int64)

In [94]:
# Compare predictions to truth labels to evaluate the model
y_preds = clf.predict(X_test)
np.mean(y_preds == y_test) # this result is the exact same number as the score method

0.8524590163934426

In [96]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_preds) # this result is also the exact same number as the score method

0.8524590163934426

In [ ]:
# Make predictions with predict_proba()